ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp

```diff
@@ -452,8 +456,118 @@ void FlightDataModel::showErrorDialog(const char* title, const char* message)
     msgBox.setStandardButtons(QMessageBox::Ok);
     msgBox.exec();
     return;
+}
+
+void FlightDataModel::pauseValidation(bool pausing) {
+    valPaused = pausing;
+
+    if (!pausing) {
+        fixupValidationErrors();
+    }
+}
+
+void FlightDataModel::fixupValidationErrors()
+{
+    struct FlightDataModel::NED prevNED;
+
+    // Skip validation when there's a download from firmware in process.
+    if (valPaused) { return; }
+
+    for (int i = 0; i < rowCount(); i++) {
+	PathPlanData * row = dataStorage.at(i);
+
+        bool dirty = false;
+
+        struct FlightDataModel::NED thisNED = getNED(row);
+
+        if (i == 0) {
+            row->mode        = Waypoint::MODE_VECTOR;
+            switch (row->mode) {
+		case Waypoint::MODE_CIRCLELEFT:
+		case Waypoint::MODE_CIRCLERIGHT:
+		    row->mode = Waypoint::MODE_VECTOR;
+
+		    showErrorDialog("Waypoint corrected",
+			    "First waypoint may not be the endpoint of an arc");
+                    dirty=true;
+
+		    break;
+		default:
+		    break;
+	    }
+
+        } else {
+	    double distance = sqrt(pow(thisNED.North - prevNED.North, 2) +
+		pow(thisNED.East - prevNED.East, 2));
+
+	    if (distance > 600) {
+		// If the distance is more than 600m, presume it's invalid.
+		// Distances larger than this begin to become problematic
+		// with local tangent plane approximation.
+
+		thisNED = prevNED;
+		setNED(row, thisNED);
+		distance = 0;
```

Comment:
I'm setting it to be the same.  This both works with the loop factoring of iterating forward over the points (deleting a point from the middle of an iterator is problematic) and allows the user to fix the waypoint.  I added a comment to this effect.

While I was just looking at it, it occurred to me that I should do this test for the first iteration of the loop, so I took all of this out of an else block.